pkg/ebpf/btf.go

```diff
@@ -22,146 +22,61 @@ import (
-		},
-	})
-
-	// 5: Ptr
-	types = append(types, &pb.BtfType{
-		NameOff:    0x0,
-		Info:       0x02000000,
-		SizeOrType: 0x4,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-
-	// 6: Func_Proto
-	types = append(types, &pb.BtfType{
-		NameOff:    0x0,
-		Info:       0x0d000002,
-		SizeOrType: 0x3,
-		Extra: &pb.BtfType_FuncProtoTypeData{
-			FuncProtoTypeData: &pb.FuncProtoTypeData{
-				Param: []*pb.BtfParam{
-					{NameOff: 0x1, ParamType: 0x3},
-					{NameOff: 0x1, ParamType: 0x5},
-				},
-			},
-		},
-	})
-
-	// 7: Func
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x0c000000,
-		SizeOrType: 0x6,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-	return &pb.TypeSection{BtfType: types}
+// SetHeaderSection takes a BTF proto and assigns the header values, magic
+// represents if the system is big or little endian (0xeb9f, 0x9feb).
+func SetHeaderSection(btf *pb.Btf, magic int32, version int32, flags int32) {
+	btf.Header = &pb.Header{
+		Magic:   magic,
+		Version: version,
+		Flags:   flags,
+		HdrLen:  0x18,
+		TypeOff: 0x0,
+		TypeLen: int32(proto.Size(btf.TypeSection)),
+		StrOff:  int32(proto.Size(btf.TypeSection)),
+		StrLen:  int32(proto.Size(btf.StringSection)),
+	}
 }
 
-func stringSection() *pb.StringSection {
-	// For buzzer's BTF Implementation the string sections is not currently
-	// in used.
-	return &pb.StringSection{Str: "buzzer"}
+// encodeTypeInfo returns the parameter type info encoded,
+// the bits arrangement is:
+//   - bits 0-15  = vlen
+//   - bits 16-23 = unused
+//   - bits 24-28 = kind
+//   - bits 29-30 = unused
+//   - bits    31 = kind_flag
+//
+// https://docs.kernel.org/bpf/btf.html#:~:text=Each%20type%20contains%20the%20following%20common%20data%3A
+func encodeTypeInfo(typeInfo *pb.TypeInfo) int32 {
+	info := int32(0)
+	info |= int32(typeInfo.Vlen)
+	info |= int32(typeInfo.Kind) << 24
+	if typeInfo.KindFlag {
+		info |= 1 << 28
+	} else {
+
+		info |= 0 << 28
+	}
+	return info
 }
 
-func getBtf() *pb.Btf {
-	// https://docs.kernel.org/bpf/btf.html#btf-type-and-string-encoding
-	type_section := typeSection()
-	string_section := stringSection()
-	header := &pb.Header{
-		Magic:   0xeb9f,
-		Version: 0x01,
-		Flags:   0x00,
-		HdrLen:  0x18,
-		TypeOff: 0x0,
-		TypeLen: int32(proto.Size(type_section)),
-		StrOff:  int32(proto.Size(type_section)),
-		StrLen:  int32(proto.Size(string_section)),
-	}
-	return &pb.Btf{
-		Header:        header,
-		TypeSection:   type_section,
-		StringSection: string_section,
+// GetBuffer takes a BTF Proto and returns its serialized value as a byte array.
+func GetBuffer(btf *pb.Btf) ([]byte, error) {
+	buffer, err := generateBtf(btf)
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
 	}
+	return buffer, nil
 }
 
-// GenerateBtf returns a byte array containing the serialized BTF data from a BTF proto.
-func GenerateBtf() ([]byte, error) {
-	btf_proto := getBtf()
+func generateBtf(btf_proto *pb.Btf) ([]byte, error) {
```

Comment:
nit: go style is to capitalize acronyms in function names (e.g. generateBTF

Reply:
Done!